### PR TITLE
Database setup

### PR DESCRIPTION
**Pull request by SafetImamovic**
The database setup includes:

- SQLAlchemy model definitions
- Pydantic models (schema) definitions
- crud.py file for different CRUD operations
- Path operations within main.py
- Seeding default records
- Simple HTML, CSS and JS frontend for API CRUD demonstration purposes

The process of running the project is outlined in the README.md

### CODE REVIEW

File: alembic/env.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 # access to the values within the .ini file in use.
 config = context.config
 
-config.set_main_option('sqlalchemy.url', os.getenv('SQL_ALCHEMY_URL'))
+config.set_main_option('sqlalchemy.url', os.getenv('ALEMBIC_DB_URL'))
```
Comment (stankovicdejan) [inline alembic/env.py]: It is not a big deal, but why did you rename this env variable to 'ALEMBIC_...'? If it is a connection string to db it is not related to alembic. Even if we were not using Alembic, the same value would be there, so some names related to db might be a better option.

File: app/crud.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        user_type_id=user.user_type_id,
+        hashed_password=hashed_password
+    )
+    db.add(db_user)
+    db.commit()
+    db.refresh(db_user)
+    return db_user
+
+
+# def delete_user(db: Session, user_id: int) -> models.User:
```
Comment (stankovicdejan) [inline app/crud.py]: Whatever we've decided to go with hard or soft delete, the other option shouldn't stay commented there in the repo. In general, commented snippets of the code are not something we would prefer to commit on dev/main branches.

File: app/crud.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+def soft_delete_user(db: Session, user_id: int):
+    db_user = db.query(models.User).filter(models.User.user_id == user_id).first()
+    if db_user:
+        db_user.is_active = False
+        db.commit()
+        db.refresh(db_user)
+    return db_user
+
+
+def reactivate_user(db: Session, user_id: int):
```
Comment (stankovicdejan) [inline app/crud.py]: This is an interesting part. If there should be an option to reactivate the user in some situations, then we don't even have 'delete' functionality, since it is temporary and might be treated as an update. In that case, we can change the methods and features to do 'deactivate_user' and 'reactivate_user' and in both cases we would go with PUT http method. 
Otherwise, if we have just one case when we can delete a user (no matter if it is a soft or hard delete) and that action cannot be reverted in any way, then it makes sense to go with the DELETE http method. Even if it is a soft delete there might be a benefit to keeping that record if there is some historical preview on the UI, or something similar. However, it would be more logical to have similar names for two functions doing two opposite actions. (like deactivate_user and reactivate_user

File: app/crud.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    db.delete(db_user_type)
+    db.commit()
+    return db_user_type
+
+
+# -------------------------------------------------------------------------------------------------
+# Course specific CRUD operations, # of functions = 5
+# -------------------------------------------------------------------------------------------------
+
+def get_course_by_id(db: Session, course_id: int) -> models.Course:
```
Comment (stankovicdejan) [inline app/crud.py]: Split functions in separate files according to the model they are supposed to modify. Imagine if we would have 30, 50 or 100 tables. Tracking all the changes in a single file would be so difficult.

File: app/crud.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+def soft_delete_course(db: Session, course_id: int):
+    db_course = db.query(models.Course).filter(models.Course.user_id == course_id).first()
+    if db_course:
+        db_course.is_active = False
+        db.commit()
+        db.refresh(db_course)
+    return db_course
+
+
+def reactivate_course(db: Session, course_id: int):
```
Comment (stankovicdejan) [inline app/crud.py]: All the same as for the user delete feature.

File: app/initial_records.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+default_users = [
```
Comment (stankovicdejan) [inline app/initial_records.py]: create a more granular folder structure for the project. If you are using an MVC pattern then split parts for model, view, and controller in separate folders and mode all the 'model' related things in that folder

File: app/main.py [hunk 7]
```diff
@@ -1,8 +1,673 @@
-from fastapi import FastAPI
```
Comment (stankovicdejan) [inline app/main.py:1]: app/main.py is an entry point of the app and should not contain all the routes and APIs implementations, but just imports and only things important for running up the app. So as mentioned before, try to implement MVC pattern and mode all these functions in matching folders and files. You can implement routers for each route according to the models you have in the project. For each route implement api calls that should be under that route. For each router, you should have a controller doing the logic. At the end of the chain, there is a model that the controller is manipulating.

File: app/models.py [hunk 8]
```diff
@@ -1,11 +1,68 @@
-import datetime
```
Comment (stankovicdejan) [inline app/models.py:1]: Again, each model should be in a separate file

**Review by stankovicdejan — COMMENTED** [review]
Please resolve all the comments.

**Comment by SafetImamovic** [comment]
New file structure

File: app/database/schemas/users.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    is_active: bool = True
+    user_type_id: int
+
+
+class UserCreate(UserBase):
+    password: str = Field(..., min_length=6)
+
+
+class UserUpdate(UserBase):
+    hashed_password: Optional[str] = Field(None, min_length=6)
```
Comment (stankovicdejan) [inline app/database/schemas/users.py]: It is pointless to use length validation on hashed_password in this way. If we have a hashed password, then the value should be the same for each value that is used as an input for the hash function. In other words, you could ensure that the length must be exactly 256, or 512, or whatever hash function is planned to be used.

File: app/main.py (hunk 7 above)
Comment (stankovicdejan) [reply app/main.py:1]: commonly, main.py, as an entry point, is stored in the root folder of the project, so, in this case, it would be 
outside the app folder, where the rest of the source code is written.

File: app/routers/courses.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+
+@router.get('/count/', response_model=int)
+def get_courses_count(db: Session = Depends(get_db)):
+    """
+    This function returns the number of courses in the database
+    :param db: The database session to use
+    :return: The number of courses in the database
+    """
+    return db.query(models.Course).count()
```
Comment (stankovicdejan) [inline app/routers/courses.py]: It would be better if we ensure that our controller for the explicit router is doing all the model manipulation. In some functions, there is 'crud.some_method()', but in some there is direct 'db.query()'.
It is not common for controllers to be named 'something_crud', and they should be dedicated to doing all the things in routers. In complex projects, they can call the service layer which will do model manipulation.

File: app/utils/seeding.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    This function seeds the user_types table with some initial values
+    :param db:
+    :return:
+    """
+    if db.query(user_types.UserType).count() == 0:
+        reset_auto_increment(db, 'user_types', 'user_type_id')
+
+        default_user_types = ['Admin', 'Student']
+
+        # results = []
```
Comment (stankovicdejan) [inline app/utils/seeding.py]: please, remove commented, unused code

**Review by stankovicdejan — COMMENTED** [review]
It is much better now, and some kind of MVC is there, but, please, check the comments below.

**Review by stankovicdejan — APPROVED** [review]
